File: game/game_logic/player.py
# Manages the player's state, including energy, resources, and companions.
from dataclasses import dataclass, field
from typing import Dict
from ..game_objects.companion import Companion
from ..game_objects.sticker import Sticker

@dataclass
class Player:
    """Represents the player and their current state in the game."""
    name: str
    energy: int = 50
    coins: int = 1000
    lumber: int = 0
    stone: int = 0
    gems: int = 500

    # Collections
    companions: Dict[str, Companion] = field(default_factory=dict)
    stickers: Dict[str, int] = field(default_factory=dict) # sticker_id -> count

    def add_companion(self, companion: Companion):
        """Adds a new companion to the player's collection."""
        if companion.id not in self.companions:
            self.companions[companion.id] = companion
            print(f"Added {companion.name} to your collection!")

    def add_sticker(self, sticker: Sticker, quantity: int = 1):
        """Adds a sticker to the player's album."""
        self.stickers[sticker.id] = self.stickers.get(sticker.id, 0) + quantity
        print(f"You got a '{sticker.name}' sticker! You now have {self.stickers[sticker.id]}.")

    def __str__(self):
        return (
            f"Player: {self.name}\n"
            f"Energy: {self.energy}, Coins: {self.coins}, Gems: {self.gems}, Lumber: {self.lumber}, Stone: {self.stone}\n"
            f"Companions: {len(self.companions)}, Stickers: {sum(self.stickers.values())}"
        )

File: unity_project/Assets/Scripts/Player.cs
using System.Collections.Generic;
using System.Text;

/// <summary>
/// Represents the player and their current state in the game.
/// This class is a direct C# conversion of the Python Player class.
/// </summary>
public class Player
{
    public string Name { get; private set; }
    public int Energy { get; set; }
    public int Coins { get; set; }
    public int Lumber { get; set; }
    public int Stone { get; set; }
    public int Gems { get; set; }

    public Dictionary<string, Companion> Companions { get; private set; }
    public Dictionary<string, int> Stickers { get; private set; }

    public Player(string name)
    {
        Name = name;
        Energy = 50;
        Coins = 1000;
        Lumber = 0;
        Stone = 0;
        Gems = 500;
        Companions = new Dictionary<string, Companion>();
        Stickers = new Dictionary<string, int>();
    }

    /// <summary>
    /// Adds a new companion to the player's collection if it's not already present.
    /// </summary>
    public void AddCompanion(Companion companion)
    {
        if (!Companions.ContainsKey(companion.id))
        {
            Companions.Add(companion.id, companion);
            // In a real Unity game, you would use a logging framework or display this in the UI.
            // For this conversion, we'll use UnityEngine.Debug.Log in the GameManager.
        }
    }

    /// <summary>
    /// Adds a sticker to the player's album, or increments the count if it already exists.
    /// </summary>
    public void AddSticker(Sticker sticker, int quantity = 1)
    {
        if (Stickers.ContainsKey(sticker.id))
        {
            Stickers[sticker.id] += quantity;
        }
        else
        {
            Stickers.Add(sticker.id, quantity);
        }
    }

    /// <summary>
    /// Provides a string representation of the player's current status.
    /// </summary>
    public override string ToString()
    {
        var sb = new StringBuilder();
        sb.AppendLine($"Player: {Name}");
        sb.AppendLine($"Energy: {Energy}, Coins: {Coins}, Gems: {Gems}, Lumber: {Lumber}, Stone: {Stone}");
        sb.AppendLine($"Companions: {Companions.Count}, Stickers: {Stickers.Values.Count}");
        return sb.ToString();
    }
}
